Add stumpless_new_param_from_string function

panosfol opened the pull request:
This function takes a string with a specific format and returns a 'param' struct.
The commit also contains error handling and tests for success and failure cases.

This PR is in regards to issue #360.

panosfol committed:
Add stumpless_new_param_from_string function

This function takes a string with a specific format
and returns a 'param' struct.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

goatshriek commented on include/private/config/locale/bg-bg.h:
```diff
@@ -280,4 +284,3 @@ L"Stumpless registration of Windows Event Log Source"
 #  define L10N_WSAGETLASTERROR_ERROR_CODE_TYPE \
 "резултатът от WSAGetLastError след неуспешното извикване"
 
-#endif /* __STUMPLESS_PRIVATE_CONFIG_LOCALE_BG_BG_H */
```
I'm not sure why the ending `#endif` statements have been dropped here. I just tested the add_l10n_string script and it does not seem to do this - how did you modify these files?

panosfol replied:
I did it through the script. Don't know why it dropped them ill check again.

goatshriek replied:
If you are able to reproduce this with the script, please share the version of Ruby and command invocation used and I will see if I can find it.

goatshriek commented on include/stumpless/error.h:
```diff
@@ -126,8 +126,13 @@ extern "C" {
+/**
+ * The format of the given string is incorrect.
+ *
+ * @since release v2.1.0
```
This should be 2.2.0, as this will be the next release

goatshriek commented on include/stumpless/param.h:
```diff
@@ -328,6 +328,34 @@ STUMPLESS_PUBLIC_FUNCTION
+ * cancelled, since it calls stumpless_new_param() which has usage of
+ * memory management functions.
+ *
+ * @since 2.1.0
```
This should be 2.2.0 as well.

goatshriek commented on src/param.c:
```diff
@@ -142,6 +142,63 @@ stumpless_new_param( const char *name, const char *value ) {
+
+  VALIDATE_ARG_NOT_NULL( string );
+
+  int i;
+  int name_len = 0, name_start = 0, value_len = 0, value_start = 0;
```
move these declarations to their own lines, instead of grouping them together.

goatshriek replied:
These should be `size_t`, as they are all used as sizes/indexes, and should not be interpreted as signed in any case.

goatshriek commented on src/windows/stumpless.def:
```diff
@@ -210,15 +210,18 @@ EXPORTS
   stumpless_get_entry_hostname                  @196
   stumpless_set_entry_hostname                  @197
-
+  stumpless_new_param_from_string               @198
```
this should be below the v2.2.0 comment, not above it

goatshriek commented on src/param.c:
```diff
@@ -142,6 +142,63 @@ stumpless_new_param( const char *name, const char *value ) {
+  int name_len = 0, name_start = 0, value_len = 0, value_start = 0;
+
+  /* Check that the characters in 'name' are allowed. */
+  for (i = 0; string[i] != '='; i++) {
```
what happens if there is no equals sign in the string? it seems like this loop will iterate past the end? Please add a test case for this possibility.

panosfol replied:
If there is no equal sign, then we iterate until ```string[i]=='\0'``` in which case we will raise an error anyway (since ```\0``` is not an acceptable character for the name) and exit with NULL. Want me to add a clause to end the loop anyway?

goatshriek replied:
As long as you have a test case demonstrating that the case is handled correctly with the correct error, no changes are necessarily needed.

goatshriek commented on src/param.c:
```diff
@@ -142,6 +142,63 @@ stumpless_new_param( const char *name, const char *value ) {
+     We add 1 to the length for the '\0' character. */
+  char value[value_len + 1];
+  
+  strncpy(name, string + name_start, name_len);
```
`memcpy` is a better choice here, as you already know the length. This may also correct the Windows builds that are currently failing.

goatshriek reviewed (changes requested):
Thank you very much for putting this together! There are a few changes below for you to make, after which I'll do another final review. Please pay particular attention to fixing the locale and Windows builds that are failing now.

goatshriek commented on test/function/param.cpp:
```diff
@@ -628,4 +628,80 @@ namespace {
+    size_t name_length = strlen("test-param-name");
+    size_t value_length = strlen("test-param-value");
+
+    param = stumpless_new_param_from_string( string );
```
The use of a variable named `string` is confusing the header check tool, which is why you see the static analysis CI failure. It believes it is a reference to the `string` type in C++. In this case, I think it would be better to rename this variable so that there isn't a name re-use, perhaps to something like `param_string`.

goatshriek commented on test/function/param.cpp:
```diff
@@ -628,4 +628,80 @@ namespace {
+    stumpless_free_all(  );
+  }
+
+  TEST( ParamFromStringTest, InvalidName ) {
```
The code coverage run noted that one of the error handling branches is not covered by these tests, where the string ends after the first loop. Please add an invalid string that covers this case; I believe an empty string will do it?

goatshriek commented on src/param.c:
```diff
@@ -142,6 +142,71 @@ stumpless_new_param( const char *name, const char *value ) {
+  name_len = i;
+  /* Declare an array to hold the 'name' string.
+     We add 1 to the length for the '\0' character. */
+  char name[name_len + 1];
```
The Windows builds appear to be failing on these array declarations (this one and line 200). You'll likely need to change these arrays to be dynamically allocated on the heap (via `alloc_mem` and subsequent `free_mem`). I believe the dynamic size of the array based on a variable, while supported in some compilers like GCC, is not universal.

You could keep the name buffer on the stack if you size it to `STUMPLESS_MAX_PARAM_NAME_LENGTH+1` and do an additional length check when scanning the name portion of the string. The value can theoretically be any size and will therefore need to be dynamic. Windows does support dynamic stack-based buffers via `_malloca` and `_freea`, but using this in a portable manner is probably more work than it's worth for this change.

panosfol replied:
I think its better I just allocate on the heap through ```alloc_mem```.

goatshriek reviewed (changes requested):
Thank you very much for addressing the previous issues quickly! There are a few more changes for you below, and some of the locale runs are still failing as well.

Were you able to identify whether the l10n string addition script has a problem in it? I'd like to resolve that if it does have something wrong with it.

panosfol commented:
> Thank you very much for addressing the previous issues quickly! There are a few more changes for you below, and some of the locale runs are still failing as well.
> 
> Were you able to identify whether the l10n string addition script has a problem in it? I'd like to resolve that if it does have something wrong with it.

It seems there was a problem from my side because I couldn't recreate it, it works perfectly fine now.

panosfol commented:
@goatshriek Also what should I do about the ```custom_analysis``` fail which is in regards of depricated features?

EDIT: I just saw that these are just warnings and the real problem was about the ```string``` you mentioned so it should be good now.

panosfol commented:
@goatshriek  it seems there was a bad push from my side ill fix it asap

panosfol commented:
@goatshriek I've fixed the memory leak, and i fixed the alphabetical order of locales, however im having a little trouble identifying what the problem with codecov is.

goatshriek commented:
The codecov issue is stating that the lines in question are never run by any of the tests. It looks like adding an empty string didn't actually reach the branch in that conditional (lines 171-172). Is this code reachable? If so, add a test with the scenario that it is intended to handle. If not, then double check that the conditional is correct, and that it is needed. You may simply be able to remove it as dead code that is handled by the if branch above it.

panosfol commented:
> The codecov issue is stating that the lines in question are never run by any of the tests. It looks like adding an empty string didn't actually reach the branch in that conditional (lines 171-172). Is this code reachable? If so, add a test with the scenario that it is intended to handle. If not, then double check that the conditional is correct, and that it is needed. You may simply be able to remove it as dead code that is handled by the if branch above it.

Yes it seems that the condition above covers the "empty string" case, since ```\0``` is not an acceptable character, therefore we raise an error before we reach the condition ```if (string[i] == '\0')```. I think its ok to remove the second ```if``` condition.
Ill keep the "empty-string" test just to cover all cases.

inquisitev commented on include/stumpless/param.h:
```diff
@@ -328,6 +328,34 @@ STUMPLESS_PUBLIC_FUNCTION
+ * cancelled, since it calls stumpless_new_param() which has usage of
+ * memory management functions.
+ *
+ * @since 2.2.0
```
```suggestion
 * @since release v2.2.0
```

goatshriek commented:
Thanks very much for putting this together and seeing it through to the end!
